Userland/Libraries/LibCore/Stream.cpp

```diff
@@ -0,0 +1,538 @@
+        flags |= O_APPEND;
+    if (has_flag(m_mode, OpenMode::Truncate))
+        flags |= O_TRUNC;
+    if (has_flag(m_mode, OpenMode::MustBeNew))
+        flags |= O_EXCL;
+    if (!has_flag(m_mode, OpenMode::KeepOnExec))
+        flags |= O_CLOEXEC;
+    if (!has_flag(m_mode, OpenMode::Nonblocking))
+        flags |= O_NONBLOCK;
+
+#ifdef __serenity__
+    int fd = ::serenity_open(filename.characters_without_null_termination(), filename.length(), flags, permissions);
+#else
+    String filename_with_null_terminator { filename };
+    int fd = ::open(filename_with_null_terminator.characters(), flags, permissions);
+#endif
+
+    if (fd < 0) {
+        return static_cast<ErrnoCode>(errno);
+    }
+
+    m_fd = fd;
+    return KSuccess;
+}
+
+bool File::is_readable() const { return has_flag(m_mode, OpenMode::Read); }
+bool File::is_writable() const { return has_flag(m_mode, OpenMode::Write); }
+
+KResultOr<size_t> File::read(Bytes buffer)
+{
+    if (!has_flag(m_mode, OpenMode::Read)) {
+        // NOTE: POSIX says that if the fd is not open for reading, the call
+        //       will return EBADF. Since we already know whether we can or
+        //       can't read the file, let's avoid a syscall.
+        return EBADF;
+    }
+
+    ssize_t rc = ::read(m_fd, buffer.data(), buffer.size());
+    if (rc < 0) {
+        return static_cast<ErrnoCode>(errno);
+    }
+
+    m_last_read_was_eof = rc == 0;
+    return rc;
+}
+
+KResultOr<size_t> File::write(ReadonlyBytes buffer)
+{
+    if (!has_flag(m_mode, OpenMode::Write)) {
+        // NOTE: Same deal as Read.
+        return EBADF;
+    }
+
+    ssize_t rc = ::write(m_fd, buffer.data(), buffer.size());
+    if (rc < 0) {
+        return static_cast<ErrnoCode>(errno);
+    }
+
+    return rc;
+}
+
+bool File::is_eof() const { return m_last_read_was_eof; }
+
+bool File::is_open() const { return m_fd >= 0; }
+
+void File::close()
+{
+    if (!is_open()) {
+        return;
+    }
+
+    // NOTE: The closing of the file can be interrupted by a signal, in which
+    // case EINTR will be returned by the close syscall. So let's try closing
+    // the file until we aren't interrupted by rude signals. :^)
+    int rc;
+    do {
+        rc = ::close(m_fd);
+    } while (rc < 0 && errno == EINTR);
+
+    VERIFY(rc == 0);
+    m_fd = -1;
+}
+
+KResultOr<off_t> File::seek(i64 offset, SeekMode mode)
+{
+    int syscall_mode;
+    switch (mode) {
+    case SeekMode::SetPosition:
+        syscall_mode = SEEK_SET;
+        break;
+    case SeekMode::FromCurrentPosition:
+        syscall_mode = SEEK_CUR;
+        break;
+    case SeekMode::FromEndPosition:
+        syscall_mode = SEEK_END;
+        break;
+    default:
+        VERIFY_NOT_REACHED();
+    }
+
+    off_t rc = lseek(m_fd, offset, syscall_mode);
+    if (rc < 0) {
+        return static_cast<ErrnoCode>(errno);
+    }
+
+    m_last_read_was_eof = false;
+    return rc;
+}
+
+KResultOr<int> Socket::create_socket(SocketDomain domain, SocketType type)
```

Comment:
This is meant as a "create socket fd" function, I guess it could be named `Socket::create_fd` or `Socket::create_socket_fd` or something. This is not exposed to `Socket` users nevertheless, just inheriters.

Reply:
I realised that a bit too late, yeah. `create_socket_fd` sounds reasonable, then.